webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/queries/NodeFactQueryProcessor.scala

```diff
@@ -176,28 +194,36 @@ class NodeFactQueryProcessor(
     val group = if (query.composition == And) GroupAnd else GroupOr
 
     // we need a better pattern matching (extensible would be better) in place of `isInstanceOf`
-    val subgroupLines = subGroupMatcher(
-      query.criteria.collect {
-        case l if l.attribute.cType.isInstanceOf[SubGroupComparator] => l
-      },
-      groupRepo
-    )
-    val ldapLines     = ldapMatcher(
-      query.criteria.collect {
-        case l if l.attribute.cType.isInstanceOf[LDAPCriterionType] => l
-      },
-      query
-    )
-    val nodeLines     = nodeFactMatcher(query.criteria.collect {
-      case l if l.attribute.nodeCriterionMatcher != UnsupportedByNodeMinimalApi => l
-    })
+    // we prefer coreNodeFact matcher on top of LDAP, since the former is quick in cache, the latter needs IO
+    val sortedLinesIO = ZIO.foldLeft(query.criteria)(CriterionLines.empty) {
+      case (lines, l) =>
+        l match {
+          // if possible, always use that one for performance reason
+          case _ if l.attribute.nodeCriterionMatcher != UnsupportedByNodeMinimalApi =>
+            lines.modify(_.coreNodeFact).using(_.appended(l)).succeed
+          case _ if l.attribute.cType.isInstanceOf[SubGroupComparator]              =>
+            lines.modify(_.subGroup).using(_.appended(l)).succeed
+          case _ if l.attribute.cType.isInstanceOf[LDAPCriterionType]               =>
+            lines.modify(_.ldap).using(_.appended(l)).succeed
+          case _                                                                    =>
+            // ??? Should not happen, perhaps log ?
+            FactQueryProcessorLoggerPure.warn(
+              s"The criterion line with attribute '${l.attribute.name}' of type ${l.attribute.cType} can't be handled. Please report that error to maintainers"
+            ) *>
+            lines.succeed
+        }
+    }
```

Comment:
so, this the place where we attribute only one matcher kind for each line, with higher priority for NodeFact matcher than for LDAP matcher.